optimizer/evaluator/evaluator.go

```diff
@@ -849,9 +851,101 @@ func trimRight(str, remstr string) string {
 }
 
 func (e *Evaluator) funcDateArith(v *ast.FuncDateArithExpr) bool {
+	// health check for date and interval
+	nodeDate := v.Date.GetValue()
+	if types.IsNil(nodeDate) {
+		v.SetValue(nil)
+		return true
+	}
+	nodeInterval := v.Interval.GetValue()
+	if types.IsNil(nodeInterval) {
+		v.SetValue(nil)
+		return true
+	}
+	// parse date
+	fieldType := mysql.TypeDate
+	var resultField *types.FieldType
+	switch x := nodeDate.(type) {
+	case mysql.Time:
+		if (x.Type == mysql.TypeDatetime) || (x.Type == mysql.TypeTimestamp) {
+			fieldType = mysql.TypeDatetime
+		}
+	case string:
+		if !mysql.IsDateFormat(x) {
+			fieldType = mysql.TypeDatetime
+		}
+	case int64:
+		if t, err := mysql.ParseTimeFromInt64(x); err == nil {
+			if (t.Type == mysql.TypeDatetime) || (t.Type == mysql.TypeTimestamp) {
+				fieldType = mysql.TypeDatetime
+			}
+		}
+	}
+	if mysql.IsClockUnit(v.Unit) {
+		fieldType = mysql.TypeDatetime
+	}
+	resultField = types.NewFieldType(fieldType)
```

Comment:
@coocood It seems to be a choice of how much compatible you would like to keep with Mysql.
I just followed the function spec in mysql docs. It would be OK to take the other force conversion way. It's simpler for both usage and implementation.